File: app/src/main/java/com/cargopull/executor_driver/application/FcmService.java
package com.cargopull.executor_driver.application;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.cargopull.executor_driver.backend.web.ApiService;
import com.google.firebase.messaging.FirebaseMessagingService;
import com.google.firebase.messaging.RemoteMessage;
import com.google.firebase.messaging.RemoteMessage.Notification;
import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;
import io.reactivex.disposables.Disposables;
import io.reactivex.schedulers.Schedulers;
import java.util.Map;
import javax.inject.Inject;

public class FcmService extends FirebaseMessagingService {

  private Observer<Map<String, String>> fcmObserver;
  @Nullable
  private ApiService apiService;
  @NonNull
  private Disposable disposable = Disposables.empty();

  @Inject
  public void setFcmObserver(@NonNull Observer<Map<String, String>> fcmObserver) {
    this.fcmObserver = fcmObserver;
  }

  @Inject
  public void setApiService(@NonNull ApiService apiService) {
    this.apiService = apiService;
  }

  @Override
  public void onNewToken(String s) {
    super.onNewToken(s);
    if (apiService == null) {
      return;
    }
    disposable = apiService.sendFcmInstanceID(s)
        .subscribeOn(Schedulers.io())
        .subscribe(() -> {
        }, Throwable::printStackTrace);
  }

  @Override
  public void onCreate() {
    super.onCreate();
    ((MainApplication) getApplication()).getAppComponent().inject(this);
  }

  @Override
  public void onDestroy() {
    super.onDestroy();
    disposable.dispose();
  }

  @Override
  public void onMessageReceived(RemoteMessage remoteMessage) {
    super.onMessageReceived(remoteMessage);
    if (remoteMessage != null) {
      Map<String, String> dataMap = remoteMessage.getData();
      Notification notification = remoteMessage.getNotification();
      if (notification != null) {
        String title = notification.getTitle();
        if (title != null) {
          dataMap.put("title", title);
        }
        String body = notification.getBody();
        if (body != null) {
          dataMap.put("body", body);
        }
      }
      fcmObserver.onNext(dataMap);
    }
  }
}

File: app/src/main/java/com/cargopull/executor_driver/application/FcmService.kt
package com.cargopull.executor_driver.application

import com.cargopull.executor_driver.R
import com.cargopull.executor_driver.backend.ringtone.RingTonePlayer
import com.cargopull.executor_driver.backend.vibro.ShakeItPlayer
import com.cargopull.executor_driver.backend.web.ApiService
import com.google.firebase.messaging.FirebaseMessagingService
import com.google.firebase.messaging.RemoteMessage
import io.reactivex.Observer
import io.reactivex.disposables.Disposables
import io.reactivex.schedulers.Schedulers

class FcmService : FirebaseMessagingService() {

    lateinit var fcmObserver: Observer<Map<String, String>>
    lateinit var apiService: ApiService
    private var disposable = Disposables.empty()
    lateinit var ringTonePlayer: RingTonePlayer
    lateinit var shakeItPlayer: ShakeItPlayer

    override fun onNewToken(s: String) {
        super.onNewToken(s)
        disposable = apiService.sendFcmInstanceID(s)
                .subscribeOn(Schedulers.io())
                .subscribe({ }, { it.printStackTrace() })
    }

    override fun onCreate() {
        super.onCreate()
        (application as MainApplication).appComponent.inject(this)
    }

    override fun onDestroy() {
        super.onDestroy()
        disposable.dispose()
    }

    override fun onMessageReceived(remoteMessage: RemoteMessage?) {
        super.onMessageReceived(remoteMessage)
        ringTonePlayer.playRingTone(R.raw.general_notify)
        shakeItPlayer.shakeIt(R.raw.general_notify_vibro)
        remoteMessage?.let {
            val dataMap = it.data
            it.notification?.let { not ->
                not.title?.let { title -> dataMap["title"] = title }
                not.body?.let { body -> dataMap["body"] = body }
            }
            fcmObserver.onNext(dataMap)
        }
    }
}
